sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -98,6 +100,142 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
+    // (SELECT * FROM tbl t1 JOIN tbl2 t2) p
+    // JOIN
+    // (SELECT * FROM tbl t3 JOIN tbl3 t4) q
+    // ON p.t1.col = q.t3.col (this is not possible)
+    // It's because there are duplicated columns in both sides of top level join and it's not
+    // possible to fully qualified the column names in condition. Therefore, query should be
+    // rewritten so that each of the outputs of child joins are aliased, so there would be a
+    // projection with aliases between top level join and scanBuilderHolder (that has pushed
+    // child joins).
+    case node @ Join(
+      PhysicalOperation(
+        leftProjections,
+        Nil,
+        leftHolder @ ScanBuilderHolder(_, _, lBuilder: SupportsPushDownJoin)
+      ),
+      PhysicalOperation(
+        rightProjections,
+        Nil,
+        rightHolder @ ScanBuilderHolder(_, _, rBuilder: SupportsPushDownJoin)
+      ),
+      joinType,
+      condition,
+    _) if conf.dataSourceV2JoinPushdown &&
+        // We do not support pushing down anything besides AttributeReference.
+        leftProjections.forall(_.isInstanceOf[AttributeReference]) &&
+        rightProjections.forall(_.isInstanceOf[AttributeReference]) &&
+        // Cross joins are not supported because they increase the amount of data.
+        condition.isDefined &&
+        // Joins on top of sampled tables are not supported
+        leftHolder.pushedSample.isEmpty &&
+        rightHolder.pushedSample.isEmpty &&
+        lBuilder.isOtherSideCompatibleForJoin(rBuilder) =>
+      val leftSideRequiredColumnNames = getRequiredColumnNames(leftProjections, leftHolder)
+      val rightSideRequiredColumnNames = getRequiredColumnNames(rightProjections, rightHolder)
+
+      // Alias the duplicated columns from left side of the join.
+      val leftSideRequiredColumnsWithAliases = leftSideRequiredColumnNames.map { name =>
+        val aliasName =
+          if (leftSideRequiredColumnNames.count(_ == name) > 1 ||
+              rightSideRequiredColumnNames.contains(name)) {
+            generateJoinOutputAlias(name)
+          } else {
+            null
+          }
+
+        new SupportsPushDownJoin.ColumnWithAlias(name, aliasName)
+      }
+
+      // Aliasing of duplicated columns in right side is done only if there are duplicates in
+      // right side only. There won't be a conflict with left side columns because they are
+      // already aliased.
+      val rightSideRequiredColumnsWithAliases = rightSideRequiredColumnNames.map { name =>
+        val aliasName =
+          if (rightSideRequiredColumnNames.count(_ == name) > 1) {
+            generateJoinOutputAlias(name)
+          } else {
+            null
+          }
+
+        new SupportsPushDownJoin.ColumnWithAlias(name, aliasName)
+      }
+
+      // Create the AttributeMap that holds (Attribute -> Attribute with up to date name) mapping.
+      val pushedJoinOutputMap = AttributeMap[Expression](
+        node.output.asInstanceOf[Seq[AttributeReference]]
```

Comment:
do we have to use the type `AttributeReference`? `Attribute` has `def withName` as well.